logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```diff
@@ -220,6 +255,72 @@ private void innerWriteEntry(DLQEntry entry) throws IOException {
     }
 
+    private void executeAgeRetentionPolicy() throws IOException {
+        if (isOldestSegmentExpired()) {
+            deleteExpiredSegments();
+        }
+    }
+
+    private boolean isOldestSegmentExpired() {
+        if (retentionTime == null) {
+            return false;
+        }
+        final Instant now = clock.instant();
+        return oldestSegmentTimestamp
+                .map(t -> t.toInstant().isBefore(now.minus(retentionTime)))
+                .orElse(false);
+    }
+
+    private void deleteExpiredSegments() throws IOException {
+        // remove all the old segments that verifies the age retention condition
+        boolean cleanNextSegment;
+        do {
+            Path beheadedSegment = oldestSegmentPath.orElseThrow(() -> new IllegalStateException("DLQ writer can't find the oldest segment to drop on path: " + queuePath));
+            final long segmentSize = Files.size(beheadedSegment);
+            this.currentQueueSize.add(-segmentSize);
+            try {
+                Files.delete(beheadedSegment);
+                logger.debug("Deleted exceeded retained age segment file {}", beheadedSegment);
+            } catch (NoSuchFileException nsfex) {
+                // the last segment was deleted by another process, maybe the reader that's cleaning consumed segments
+                logger.info("File not found {}, maybe removed by the reader pipeline", beheadedSegment);
+            }
+            updateOldestSegmentReference();
+            cleanNextSegment = isOldestSegmentExpired();
+        } while (cleanNextSegment);
+    }
+
+    private void updateOldestSegmentReference() throws IOException {
+        oldestSegmentPath = getSegmentPaths(this.queuePath).sorted().findFirst();
+        if (!oldestSegmentPath.isPresent()) {
+            oldestSegmentTimestamp = Optional.empty();
+            return;
+        }
+        // extract the newest timestamp from the oldest segment
+        oldestSegmentTimestamp = Optional.of(readTimestampOfLastEventInSegment(oldestSegmentPath.get()));
+    }
+
+    /**
+     * Extract the timestamp from the last DLQEntry it finds in the given segment.
+     * Start from the end of the latest block, and going backward try to read the next event from its start.
+     * */
+    private static Timestamp readTimestampOfLastEventInSegment(Path segmentPath) throws IOException {
+        final int lastBlockId = (int) Math.ceil(((Files.size(segmentPath) - VERSION_SIZE) / (double) BLOCK_SIZE)) - 1;
+        byte[] eventBytes;
+        try (RecordIOReader recordReader = new RecordIOReader(segmentPath)) {
+            int blockId = lastBlockId;
+            do {
+                recordReader.seekToBlock(blockId);
+                eventBytes = recordReader.readEvent();
+                blockId--;
+            } while (eventBytes == null && blockId >= 0); // no event present in last block, try with the one before
+        }
+        if (eventBytes == null) {
```

Comment:
Right, handled in 5e8e1e56882482fdcd7a7af71d147e51d869aaea, removed the throw and handled empty optional. 
Also avoid to throw exception in 
```java
Path beheadedSegment = oldestSegmentPath.orElseThrow(() -> new IllegalStateException("DLQ writer can't find the oldest segment to drop on path: " + queuePath));
```
because the `oldestSegmentPath` could be set to empty by `updateOldestSegmentReference`method  if the last segment doesn't provide a valid timestamp